src/vos/vos_io.c

```diff
@@ -1868,16 +1896,24 @@ dkey_update(struct vos_io_context *ioc, uint32_t pm_ver, daos_key_t *dkey,
 	daos_handle_t		 ak_toh;
 	struct vos_krec_df	*krec;
 	uint32_t		 update_cond = 0;
+	uint32_t                 flags         = SUBTR_CREATE;
 	bool			 subtr_created = false;
 	int			 i, rc;
 
 	rc = obj_tree_init(obj);
 	if (rc != 0)
 		return rc;
 
-	rc = key_tree_prepare(obj, obj->obj_toh, VOS_BTR_DKEY, dkey,
-			      SUBTR_CREATE, DAOS_INTENT_UPDATE, &krec, &ak_toh,
-			      ioc->ic_ts_set);
+	if (vos_obj_flat_kv_supported(ioc->ic_cont, obj->obj_id)) {
+		if (ioc->ic_iod_nr == 1) {
```

Comment:
should the flag depend on the request?  what will happen if the array object being accessed by daos_object_xxx with multiple iod?

Reply:
In vos_update_begin, vos_check_akeys will fail the update if there is more than one iod for the same akey.   This is a simplifying assumption for things like conditional operations.

So for no akey case, this should never happen.  A flat dkey should always have 1 entry.

Perhaps, though, we should check this in vos_check_akeys.